phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexRebuildRegionScanner.java

```diff
@@ -1092,6 +1093,181 @@ private boolean isColumnIncluded(Cell cell) {
+                    put.getFamilyCellMap().remove(family);
+                }
+                return;
+            }
+        }
+    }
+
+    public static void merge(Put current, Put previous) throws IOException {
+        for (List<Cell> cells : previous.getFamilyCellMap().values()) {
+            for (Cell cell : cells) {
+                if (!current.has(CellUtil.cloneFamily(cell), CellUtil.cloneQualifier(cell))) {
+                    current.add(cell);
+                }
+            }
+        }
+    }
+
+    public static Put mergeNew(Put current, Put previous) throws IOException {
+        Put next = new Put(current);
+        merge(next, previous);
+        return next;
+    }
+
+    /**
+     * Generate the index update for a data row from the mutation that are obtained by merging the previous data row
+     * state with the pending row mutation for index rebuild. This method is called only for global indexes.
+     * pendingMutations is a sorted list of data table mutations that are used to replay index table mutations.
+     * This list is sorted in ascending order by the tuple of row key, timestamp and mutation type where delete comes
+     * after put.
+     */
+    public static List<Mutation> prepareIndexMutationsForRebuild(IndexMaintainer indexMaintainer,
+                                                                 Put dataPut, Delete dataDel) throws IOException {
+        List<Mutation> dataMutations = getMutationsWithSameTS(dataPut, dataDel);
+        List<Mutation> indexMutations = Lists.newArrayListWithExpectedSize(dataMutations.size());
+        // The row key ptr of the data table row for which we will build index rows here
+        ImmutableBytesPtr rowKeyPtr = (dataPut != null) ? new ImmutableBytesPtr(dataPut.getRow()) :
+                new ImmutableBytesPtr(dataDel.getRow());
+        // Start with empty data table row
+        Put currentDataRow = null;
+        // The index row key corresponding to the current data row
+        byte[] IndexRowKeyForCurrentDataRow = null;
+        for (Mutation mutation : dataMutations) {
+            ValueGetter currentDataRowVG = (currentDataRow == null) ? null : new IndexRebuildRegionScanner.SimpleValueGetter(currentDataRow);
+            long ts = getTimestamp(mutation);
+            if (mutation instanceof Put) {
+                // Add this put on top of the current data row state to get the next data row state
+                Put nextDataRow = (currentDataRow == null) ? new Put((Put)mutation) : mergeNew((Put)mutation, currentDataRow);
+                ValueGetter nextDataRowVG = new IndexRebuildRegionScanner.SimpleValueGetter(nextDataRow);
+                Put indexPut = prepareIndexPut(indexMaintainer, rowKeyPtr, nextDataRowVG, ts, true, null, null);
+                indexMutations.add(indexPut);
+                // Delete the current index row if the new index key is different than the current one
+                if (IndexRowKeyForCurrentDataRow != null) {
```

Comment:
nit: java variable names start with lower case.

Reply:
I am not sure how I got the upper case variable name here, probably it happened when I did bulk rename. I will change it.